### ISSUE TITLE
Non-RAW files don't get their embedded ICC profile initialized when history is deleted/compressed

### ISSUE THREAD

**Issue by aurelienpierre** [darktable-org/darktable#5291]
1. Open a JPG, TIFF, PNG, etc. file
2. Either: 
   * In darkroom, click on the 0th step in history, then compress it,
   * In lighttable, click on "discard history"
3. See the input profile set as a matrice instead of embedded ICC. Sometimes, the setting says "embedded ICC" but the image is too bright, which means the actual profile used in the pipe is linear (so, the matrice) and the actual embedded profile is not properly initialized.
4. Reset the input profile module, now the right profile (embedded ICC) is used.

## What happens ?

Input profile, as white balance, initializes its defaults at runtime, in `reload_defaults()` by reading the image storage (`module->dev->image_storage`) and extracting relevant tags (color profile or white balance coefficients). But input profile actually reads the image buffer, extracts the embedded profile, and **writes** the profile in the image storage too, meaning it is not initialized before it runs. And this is the core issue of the problem (since you may notice white balance doesn't suffer from the same issue).

When history is manipulated, it uses only the database info. No image storage is passed to history management so the defaults cannot be initialized with the relevant settings and fallback to safety defaults.

## Solution 

Option 1:
History handling functions should also care about image storage and give modules the opportunity to properly read them to init sane defaults.

Option 2:
Embedded profile extraction should be done prior to pipeline startup, so the image container is complete and consistent before launching the pipeline and history handling (and not write it in colorin `reload_defauts`). That could be when opening the image or initializing the cache container.

**Comment by jenshannoschwalm** [comment]
@aurelienpierre i guess you are very busy with other stuff atm, i have worked on both parts of the code so i think i can do that. ok?

The history compression has been done this way for performance reasons, may be we can use a img->flag that tells the image needs re-development after compression so avoiding that extra work for raws.  One reason could be an embedded icc.

**Comment by jenshannoschwalm** [comment]
I looked into this and confess: I simply can't reproduce.

Also, i doubt your analysis is fully correct.

Whenever an image `dt_image_t` is included in the cache it first has it's profile = NULL, this is populated via `reload_defaults` in colorin.c and this is kept forever until the `dt_image_t` struct is removed from the cache.

> But input profile actually reads the image buffer, extracts the embedded profile, and writes the profile in the image storage too

From colorin.c --- it reads from the file and writes the profile pointer & size into the struct and this is fully correct i think.

```dt_imageio_jpeg_t jpg;
      if(!dt_imageio_jpeg_read_header(filename, &jpg))
      {
        img->profile_size = dt_imageio_jpeg_read_profile(&jpg, &img->profile);
        use_eprofile = (img->profile_size > 0);
      }
```
Any idea? Could any of the icc reading function be wrong?

**Comment by jenshannoschwalm** [comment]
@aurelienpierre i have again tried to reproduce without "success". Could you give more details how to reproduce or is the issue in-consistant? I would love to understand and tackle ...

**Comment by TurboGit** [comment]
I cannot reproduce this with current master.

**Comment by parafin** [comment]
But you can with 3.2.1? Or is this different from #6202?

**Comment by TurboGit** [comment]
@parafin : did not test with 3.2.1 as I don't have a build of it at the moment.

In fact the issue I have reported with the image full red on display was probably a different issue and due to wrong reload_default() in demosaic (PR #6236.).

**Comment by parafin** [comment]
Please see https://github.com/darktable-org/darktable/issues/6202#issuecomment-687573271 for reproduction steps. I've tested on 3.2.1, if you can't reproduce on master I can re-check on master then.

**Comment by TurboGit** [comment]
@parafin : I confirm that I cannot reproduce on my side with option set to use embedded jepg (BTW, I cannot reproduce with the setting inverted too).

To test I have used a duplicate, discarded history of all images. Open first RAW, then a jpeg and then the duplicate of the first RAW. The rendering is exactly the same. The colorin module is reset correctly... Or maybe I'm not looking at the right place?

**Comment by parafin** [comment]
Does discard history operation reset to the freshly imported state or does it re-create the default stack? I've tested with freshly imported images without xmp files.

**Comment by TurboGit** [comment]
I have removed all XMP, reimported and still cannot reproduce with current master. Maybe this is due to SOME jpeg having a specific embedded profile?

**Comment by parafin** [comment]
No, it's just default sRGB. I'll re-test on master.

**Comment by jenshannoschwalm** [comment]
I also could not reproduce with @parafin suggestion.
I used these, put them in a empty directory and imported. Discarding history on all three images, open jpeg first and switch via filmstrip or pressing space and voila, tested on master this early morning

https://www.amazon.com/clouddrive/share/apaZI8JqnaVzyu4PC1ryo3tRvKX2AtJsbY7W5keegCf

**Comment by jenshannoschwalm** [comment]
Added some printfs into reload_defaults in colorin and rawprepare iop's, for me it was surprising i did not see them executed while switching on filmstrip. ???

**Comment by TurboGit** [comment]
Ok, I can confirm with these set of image. And indeed this is because the JPEG has a profile embedded. I'll see if I can make sense of this...

**Comment by TurboGit** [comment]
Ok, see #6238.

**Comment by TurboGit** [comment]
Note that this is the same kind of fix I made this morning for demosaic in #6236.
